compiler/rustc_monomorphize/src/collector/abi_check.rs

```diff
@@ -0,0 +1,97 @@
+use rustc_abi::Abi;
+use rustc_middle::ty::{self, Instance, InstanceKind, ParamEnv, Ty, TyCtxt};
+use rustc_span::def_id::DefId;
+use rustc_span::{Span, Symbol};
+use rustc_target::abi::call::{FnAbi, PassMode};
+
+use crate::errors::{AbiErrorDisabledVectorTypeCall, AbiErrorDisabledVectorTypeDef};
+
+// Represents the least-constraining feature that is required for vector types up to a certain size
+// to have their "proper" ABI.
+const X86_VECTOR_FEATURES: &'static [(u64, &'static str)] =
+    &[(128, "sse"), (256, "avx"), (512, "avx512f")];
+
+fn do_check_abi<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    abi: &FnAbi<'tcx, Ty<'tcx>>,
+    target_feature_def: DefId,
+    emit_err: impl Fn(&'static str),
+) {
+    let feature_def = if tcx.sess.target.arch == "x86" || tcx.sess.target.arch == "x86_64" {
+        X86_VECTOR_FEATURES
+    } else if tcx.sess.target.arch == "aarch64" {
+        // ABI on aarch64 does not depend on target features.
+        return;
+    } else {
+        // FIXME: add support for non-tier1 architectures
+        return;
+    };
+    let codegen_attrs = tcx.codegen_fn_attrs(target_feature_def);
+    for arg_abi in abi.args.iter().chain(std::iter::once(&abi.ret)) {
+        let size = arg_abi.layout.size;
+        if matches!(arg_abi.layout.abi, Abi::Vector { .. })
+            && !matches!(arg_abi.mode, PassMode::Indirect { .. })
+        {
+            let feature = match feature_def.iter().find(|(bits, _)| size.bits() <= *bits) {
+                Some((_, feature)) => feature,
+                None => panic!("Unknown vector size: {}; arg = {:?}", size.bits(), arg_abi),
+            };
+            let feature_sym = Symbol::intern(feature);
+            if !tcx.sess.unstable_target_features.contains(&feature_sym)
+                && !codegen_attrs.target_features.iter().any(|x| x.name == feature_sym)
```

Comment:
`unstable_target_features` is an odd name... just leaving a note here so that the t-compiler reviewer can carefully check that this is checking the right thing: we have to ensure that the target feature is enabled when generating code for this function.